vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1259 @@
+                node.second->on_route = true;
+                root_node = node.second;
+            }
+            if (used_by_pin_.find(std::tuple<int, loc_>{node_id, node.second->loc}) != used_by_pin_.end()) {
+                node.second->has_pins = true;
+                node.second->on_route = true;
+                root_node = node.second;
+            }
+        }
+        return root_node;
+    }
+
+    /*
+     * Removes dangling nodes from a graph represented by the root node.
+     * Dangling nodes are nodes that do not connect to a pin, a pip or other non-dangling node.
+     */
+
+    int reduce_graph_and_count_nodes(node_* root) {
+        int cnt = 0;
+        std::queue<node_*> walker;
+        std::stack<node_*> back_walker;
+        walker.push(root);
+        root->visited = root;
+        bool all_nulls;
+        while (!walker.empty()) {
+            all_nulls = true;
+            node_* vertex = walker.front();
+            walker.pop();
+            back_walker.push(vertex);
+            for (int i = 0; i < 4; ++i) {
+                if (vertex->links[i] != nullptr) {
+                    all_nulls = false;
+                    if (i == 0 || i == 1)
+                        cnt++;
+                    if (vertex->links[i]->visited == nullptr) {
+                        vertex->links[i]->visited = vertex;
+                        walker.push(vertex->links[i]);
+                    }
+                }
+            }
+            cnt += all_nulls;
+        }
+        while (!back_walker.empty()) {
+            node_* vertex = back_walker.top();
+            back_walker.pop();
+            if (!vertex->has_pins && !vertex->on_route) {
+                if (vertex->visited->segment_id == -1) {
+                    vertex->visited->segment_id = vertex->segment_id;
+                }
+                for (int i = 0; i < 4; ++i) {
+                    node_* temp = nullptr;
+                    if (vertex->links[i] != nullptr) {
+                        if (i == 2 || i == 3)
+                            cnt--;
+                        temp = vertex->links[i];
+                        temp->links[(i + 2) % 4] = nullptr;
+                    }
+                }
+                delete (vertex);
+            } else if (vertex->on_route) {
+                vertex->visited->on_route = true;
+                if (vertex->visited->segment_id == -1)
+                    vertex->visited->segment_id = vertex->segment_id;
+            }
+        }
+        return cnt;
+    }
+
+    void populate_CHAN_loc_map(int node_id, node_* root, float R, float C) {
+        /*
+         * Create future rr_nodes that corespond to given FPGA Interchange node
+         */
+
+        std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, int /*index*/> loc_chan_idx;
+
+        std::queue<node_*> walker;
+        std::stack<node_*> back_walker;
+        walker.push(root);
+        bool all_nulls, CHANX, CHANY;
```

Comment:
```suggestion
        bool all_nulls, is_chanx, is_chany;
```

To better understand that this is a boolean varialbe it is good to add some more descriptive name